testsuite/tests/typing-poly/poly_params.ml

```diff
@@ -0,0 +1,202 @@
+Error: This argument has type int -> int which is less general than
+         'a. 'a -> 'a
+|}];;
+
+let rec poly5 : bool -> ('a. 'a -> 'a) -> int * string =
+  fun p id ->
+    if p then poly5 false id else id 5, id "five"
+[%%expect {|
+val poly5 : bool -> ('a. 'a -> 'a) -> int * string = <fun>
+|}];;
+
+let _ = poly5 true (fun x -> x)
+[%%expect {|
+- : int * string = (5, "five")
+|}];;
+
+let _ = poly5 true (fun x -> x + 1)
+[%%expect {|
+Line 1, characters 19-35:
+1 | let _ = poly5 true (fun x -> x + 1)
+                       ^^^^^^^^^^^^^^^^
+Error: This argument has type int -> int which is less general than
+         'a. 'a -> 'a
+|}];;
+
+
+let rec poly6 : 'b. bool -> ('a. 'a -> 'a) -> 'b -> 'b * 'b option =
+  fun p id x ->
+    if p then poly6 false id x else id x, id (Some x)
+[%%expect {|
+val poly6 : bool -> ('a. 'a -> 'a) -> 'b -> 'b * 'b option = <fun>
+|}];;
+
+let _ = poly6 true (fun x -> x) 8
+[%%expect {|
+- : int * int option = (8, Some 8)
+|}];;
+
+let _ = poly6 true (fun x -> x + 1) 8
+[%%expect {|
+Line 1, characters 19-35:
+1 | let _ = poly6 true (fun x -> x + 1) 8
+                       ^^^^^^^^^^^^^^^^
+Error: This argument has type int -> int which is less general than
+         'a. 'a -> 'a
+|}];;
+
+let needs_magic (magic : 'a 'b. 'a -> 'b) = (magic 5 : string)
+let _ = needs_magic (fun x -> x)
+[%%expect {|
+val needs_magic : ('a 'b. 'a -> 'b) -> string = <fun>
+Line 2, characters 20-32:
+2 | let _ = needs_magic (fun x -> x)
+                        ^^^^^^^^^^^^
+Error: This argument has type 'c. 'c -> 'c which is less general than
+         'a 'b. 'a -> 'b
+|}];;
+
+let with_id (f : ('a. 'a -> 'a) -> 'b) = f (fun x -> x)
+[%%expect {|
+val with_id : (('a. 'a -> 'a) -> 'b) -> 'b = <fun>
+|}];;
+
+let _ = with_id (fun id -> id 4, id "four")
+[%%expect {|
+- : int * string = (4, "four")
+|}];;
+
+let non_principal1 p f =
+  if p then with_id f
+  else f (fun x -> x)
+[%%expect {|
+val non_principal1 : bool -> (('a. 'a -> 'a) -> 'b) -> 'b = <fun>
+|}, Principal{|
+Line 3, characters 7-21:
+3 |   else f (fun x -> x)
+           ^^^^^^^^^^^^^^
+Warning 18 [not-principal]: applying a higher-rank function here is not principal.
+val non_principal1 : bool -> (('a. 'a -> 'a) -> 'b) -> 'b = <fun>
+|}];;
+
+let non_principal2 p f =
+  if p then f (fun x -> x)
+  else with_id f
+[%%expect {|
+Line 3, characters 15-16:
+3 |   else with_id f
+                   ^
+Error: This expression has type ('b -> 'b) -> 'c
+       but an expression was expected of type ('a. 'a -> 'a) -> 'd
+       The universal variable 'a would escape its scope
+|}];;
+
+let principal1 p (f : ('a. 'a -> 'a) -> 'b) =
+  if p then f (fun x -> x)
+  else with_id f
+[%%expect {|
+val principal1 : bool -> (('a. 'a -> 'a) -> 'b) -> 'b = <fun>
+|}];;
+
+let principal2 : bool -> (('a. 'a -> 'a) -> 'b) -> 'b =
+  fun p f ->
+    if p then f (fun x -> x)
+    else with_id f
+[%%expect {|
+val principal2 : bool -> (('a. 'a -> 'a) -> 'b) -> 'b = <fun>
+|}];;
```

Comment:
I have experimented and would like to add some more tests.

```
type poly = ('a. 'a -> 'a) -> int * string   (* works fine *)
type 'a arg = 'b
  constraint 'a = 'b -> 'c
type really_poly = poly arg  (* errors [good!] but with poor error message *)
```

But beyond just this strange case, I also have observed that `poly option` is a thing, and it seems to work. I can't quite work out the consequences of this, though. That is, the system you've designed has limited [guarded impredicativity](https://dl.acm.org/doi/abs/10.1145/3192366.3192389) (where `->` is invariant). Is this intentional?

This design certainly leads to more principality trouble. Here are some tests:

```
let polys : poly option list = [ None; Some (fun x -> x 5, x "hello") ]

let take_polys : poly option list -> int = fun x -> List.length x

let test_take_polys = take_polys [ Some (fun x -> x 5, x "hello") ]

let test_no_take_polys = List.length ([ Some (fun x -> x 5, x "hello") ] : poly option list)

let test_no_take_polys2 = List.length [ (Some (fun x -> x 5, x "hello") : poly option) ]
let test_no_take_polys3 = List.length [ (Some (fun x -> x 5, x "hello") : poly option); Some (fun y -> y 6, y "goodbye") ]

let test_no_take_polys4 = List.length [ Some (fun y -> y 6, y "goodbye"); (Some (fun x -> x 5, x "hello") : poly option) ]
```

All of these are accepted except the last, which is identical to the penultimate, except that the elements in the list are reordered. I can't quite say this is deep trouble, but I think we would have a hard time writing down a specification of what programs are accepted and what are rejected, without modeling the order that type inference examines arguments (which, incidentally, is *backward* to the order that arguments are evaluated). 

Somewhat separately, it's worth having tests confirming that `->` is invariant (which is a good design decision, in my opinion):

```
type p1 = ('a. 'a -> 'a) -> int
type p2 = (bool -> bool) -> int

let x1 : p1 = fun _ -> 1
let x2 : p2 = fun _ -> 2

let x1_2 : p1 = x2
let x1_2' : p1 = fun z -> x2 z
```

`x1_2` is rejected while `x1_2'` is accepted, because the eta-expansion is necessary due to `->`'s invariance.

-----------------------------------

The more I think about this, the less I like it. Maybe it's just because I'm so used to principality. But I worry that anything non-principal is a bug factory. Non-principality means that whether or not a program is accepted depends on the vagaries of type inference. The tests above have only started to explore the possibilities. For example, we haven't touched on the possibility of polytype synonyms in modules. Or in first-class modules. Or in hypothetical implicit first-class modules. Maybe it's all OK, but the whole inference algorithm here depends on knowing the shape of a type early on in inference, and complicated scenarios might arise (now or in the future) where we can't know that shape until after some solving is done. A thorny case is GADTs:

```
type ('a, 'b) eq =
  | Refl : ('a, 'a) eq

type 'a exeq =
  | Ex : 'b * ('b, 'a) eq -> 'a exeq

let f : (poly -> unit) exeq -> unit = function
  | Ex (p, Refl) -> p (fun x -> x 5, x "string")
```

To my surprise, this is accepted. (That's OK -- it's sound.) But a lot of steam escaped ocaml's ears to accept this. Critically, ocaml had to know `p`'s type before checking the application of `p` -- even though `p`'s type can be known only by looking at GADT-pattern-match-introduced equalities. GHC cannot do this and does not attempt it. The key question to me is: are we confident that all future ocamls will be as enterprising? I doubt it. (That's the nature of non-principality.) And so by allowing a definition like `f` to be accepted, we're making an implicit commitment of future support that may come back to bite us.

Admittedly, we're in a different environment than GHC: the code we care most about is within our own walls. We could, for example, build all of that code on a future ocaml to find out if we've reneged on any promises. Yet relying on that ability when designing a new feature makes me unhappy.

Reply:
Did you try adding these tests to the testsuite, or run them by hand? If you ran them by hand, did you also try with `-principal` turned on?

I suspect that you would have got a principality warning for `test_no_take_polys3` with `-principal` enabled. Similarly, any other examples breaking principality should warn with that option enabled.

Reply:
I've put the typing rules for the `-principal` version here:

https://gist.github.com/lpw25/fa04a679d186325c0f2025fc2c495f3b

They are a variation of the rules from [Extending ML with Semi-Explicit Higher-order Polymorphism](http://gallium.inria.fr/~remy/ftp/tacs97.pdf).

Technically, I've implemented something slightly more expressive than those rules, because I allow:
```ocaml
let foo : ('a. 'a -> 'a) -> int * string = fun id -> id 4, id "four"
```
whilst those rules require the actual parameter to be annotated:
```ocaml
let foo : ('a. 'a -> 'a) -> int * string = fun (id : 'a. 'a -> 'a) -> id 4, id "four"
```
I could probably write some rules for this minor extension, but unlike the rules I wrote above I couldn't just base them on an existing paper, so it would be more work.

Reply:
Note that I don't consider this system to have guarded impredicative polymorphism. That's really about allowing a universal variable to be instantiated with a type scheme directly, which is very difficult to support. Whereas all our type scheme's are part of a larger construct -- polymorphic parameters are an inseparable part of the function type, the existing polymorphic methods are an inseparable part of the enclosing object type, etc. That means that we can't even express the difficult cases.

Reply:
I've added versions of your various tests to the testsuite.